v0

Fakrichen opened the pull request:
version 0 of the solution.

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,202 @@
+        return x0_->value();
+    }
+
+    Real ConstantBlackScholesProcess::drift(Time t, Real x) const {
```
You can simplify a lot more than this.

The idea is that, since you know the exercise date and the strike (that you passed to the constructor) you can extract the corresponding volatility and zero rates just once, store them as precalculated numbers, and use them in these methods directly without making any other calls to `zeroRate` (here) or `blackVolatility` (in diffusion).

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,202 @@
+    }
+
+    Real ConstantBlackScholesProcess::diffusion(Time t, Real x) const {
+        return blackVolatility_;
```
This can't compile.  `blackVolatility_` is a whole term structure.  The method needs to return a number.

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,202 @@
+
+    const Handle<YieldTermStructure>&
+    ConstantBlackScholesProcess::dividendYield() const {
+        return dividendYield_->zeroRate(t0, Continuous, NoFrequency, true) ;
```
This can't compile either.  You're returning a rate instead of the `Handle<YieldTermStructure>` that you declare.

Anyway, you don't need these methods (i.e., the ones returning handles) for the purpose of this exercise.  They won't be used by the engine.

lballabio commented on project1/READ ME.txt:
```diff
@@ -0,0 +1,20 @@
+
+
+Question:
+are we supposed to consider the case of Strike-dependent Volatility ? 
```
No, the volatility is not strike-dependent.

lballabio commented on project1/READ ME.txt:
```diff
@@ -0,0 +1,20 @@
+
+Question:
+are we supposed to consider the case of Strike-dependent Volatility ? 
+In our case (constant volatility) the black-volatility is equal to the local volatility?
```
It is.  You can ignore the local volatility.

lballabio commented on project1/READ ME.txt:
```diff
@@ -0,0 +1,20 @@
+Question:
+are we supposed to consider the case of Strike-dependent Volatility ? 
+In our case (constant volatility) the black-volatility is equal to the local volatility?
+zeroRate (time t, arg) = zeroRate (Date & d, Date d1 , args ) ? what is the difference? 
```
You can see a particular date as the date itself (the case `zeroRate(d, ...)` or as a time, expressed as fraction of year, from the reference date of the curve; e.g., 6 months from the reference date is 0.5 (or something close to 0.5, depending on the day counter), 3 months is about 0.25.  That would be the case `zeroRate(t, ...)`.  For convenience, the curve provides both methods.

lballabio commented on project1/READ ME.txt:
```diff
@@ -0,0 +1,20 @@
+are we supposed to consider the case of Strike-dependent Volatility ? 
+In our case (constant volatility) the black-volatility is equal to the local volatility?
+zeroRate (time t, arg) = zeroRate (Date & d, Date d1 , args ) ? what is the difference? 
+Are we supposed to keep all the specific models (GarmanKohlagenProcess, BlackProcess ..) ? 
```
No, not really.  If you're only using the process inside the engine, you can use the generic one.

lballabio commented on project1/READ ME.txt:
```diff
@@ -0,0 +1,20 @@
+In our case (constant volatility) the black-volatility is equal to the local volatility?
+zeroRate (time t, arg) = zeroRate (Date & d, Date d1 , args ) ? what is the difference? 
+Are we supposed to keep all the specific models (GarmanKohlagenProcess, BlackProcess ..) ? 
+what does the RegisterWith do ? 
```
It sets up the process to receive notifications when any of the curves change.  (See the part in the book that describes the observer/observable machinery.)

lballabio commented:
Folks, how are things going?  Do you need any help?

Fakrichen commented:
Hello sir,
We have corrected the errors mentioned on your comments and updated the version of ConstantBlackScholesProcess. 
Besides, we are trying to compile the code in order to test it and compare it to BlackScholesProcess.
Would you please give us some comments on our  latest version of the code?

Thank you very much

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,76 @@
+* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
```
```suggestion
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
```
(Somehow the first `/` was lost and this was no longer a comment.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,13 +1,81 @@
+		Real x0() const;
+		Real drift(Time t, Real s) const;
+		Real diffusion(Time t, Real s) const;
+		Real variance(const StochasticProcess1D&,
```
The `variance` method looks like something copied by mistake from the `discretization` interface. You can remove it (both here and in the cpp file).

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -26,10 +26,15 @@
 #include <ql/processes/blackscholesprocess.hpp>
 #include <ql/termstructures/volatility/equityfx/blackconstantvol.hpp>
 #include <ql/termstructures/volatility/equityfx/blackvariancecurve.hpp>
+#include <ql/quantlib.hpp>
```
I don't think you need to import all `quantlib.hpp`.  The headers you imported above should be enough.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,13 +1,81 @@
+                 the internal calculations work on \f$ ln S \f$.
+        \ingroup processes
+    */
+   namespace QuantLib {
```
You're already inside namespace QuantLib (See line 33).

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,13 +1,81 @@
+    \brief Black-Scholes processes
+*/
+
+#ifndef quantlib_black_scholes_process_hpp
```
You need to change this one (which is already defined in the original `blackscholesprocess.hpp`) and use a unique one, otherwise this header will be skipped by the preprocessor.

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -60,10 +67,72 @@ namespace QuantLib {
+                                                                process->dividendYield(),
+                                                                boost::shared_ptr<StochasticProcess1D::discretization>(new EulerDiscretization)));*/
+                                //
+                boost::shared_ptr<StochasticProcess1D::discretization> disc = new EulerDiscretization ;
```
```suggestion
                boost::shared_ptr<StochasticProcess1D::discretization> disc(new EulerDiscretization);
```
(some compilers don't accept your way of writing it).
An alternative would be to give a default discretization to the `constantBlackScholesProcess` ( see the comment on this in the other header).  This way, you won't even need to pass it here.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,13 +1,81 @@
+			const Handle<YieldTermStructure>& risk_free_BS,
+			const Handle<BlackVolTermStructure>& volatility_BS,
+			const Handle<YieldTermStructure>& dividend_yield_BS,
+			boost::shared_ptr<discretization>& disc);
```
```suggestion
			const boost::shared_ptr<discretization>& disc = boost::shared_ptr<discretization>(new EulerDiscretization));
```
This way, you're not forced to pass a discretization if this default is ok.

salma26 commented:
Problem : 
We have a problem while executing the main, the error message is Boost Assertion failed : px != 0.
I figuered out that the program doesn't enter to the pathGenerator method. 
Could you please check the code ? (it compiles successfully)

lballabio commented:
Ok, I'll have a look.

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -60,10 +66,58 @@ namespace QuantLib {
+       
+      private: 
+        bool constant_; //! définition de l'attribut boolean
+        boost::shared_ptr<GeneralizedBlackScholesProcess> process_;
```
The problem is here.  You're redeclaring the data members from the base class.  What happens is that the constructor initializes the `process_` data member from the base class, and the code in `pathGenerator` uses `process_` from this class (which was not initialized).  You need to remove all data members here except for `constant_`.  If the compiler complains, you might need to replace some `process_` with `this->process_` (the same goes for other data members).

Fakrichen commented:
Hello Sir,
thank you for your support.
We succesfully executed the code.
Please find our results in the final report "QuantLib.pdf".
Regards,
Project1 team
